helix-rest/src/main/java/org/apache/helix/rest/server/HelixRestServer.java

```diff
@@ -175,6 +197,8 @@ protected ResourceConfig getResourceConfig(HelixRestNamespace namespace, Servlet
       cfg.register(new CORSFilter());
     }
     cfg.register(new AuditLogFilter(_auditLoggers));
+    cfg.register(new ClusterAuthFilter(_clusterAuthValidator));
+    cfg.register(new NamespaceAuthFilter(_namespaceAuthValidator));
```

Comment:
Is there a strict hierarchy/precedence relationship between cluster and namespace or are they independent?

Reply:
They are independent: someone could have access to a cluster but not the namespace, vice versa.

Reply:
That gets me to wonder why we're not using both annotations for each endpoint? 

Suppose Alice has access to the cluster but not the namespace. Do we need to fail Alice's requests because in Helix, there is a hierarchy between namespaces and clusters - a namespace can have one or more clusters, and clusters must belong to a namespace (unless namespace isn't enabled)?

Reply:
If Alice has access to the particular cluster but not the namespace, then all requests related to that cluster will succeed; but Alice is not allowed to create clusters or delete clusters. These are 2 different types of privilege. If Alice has access to the namespace, it means Alice can create/delete clusters, but Alice cannot necessarily access a cluster created by someone else (ClusterAuth). 

In the approach you proposed, there needs to be a way to differentiate endpoints based on ClusterAuth and endpoints based on NamespaceAuth, meaning that there must be a list somewhere, hardcoded or configured. I prefer having different annotations to mark which endpoint needs ClusterAuth/NamespaceAuth, than using a list to identify them; the annotations is a cleaner approach to me.

Reply:
Thanks for explaining, it makes sense. I think the current approach should work.